File: cmd/open/repo.go
package open

import (
	"github.com/fioncat/roxide/cmd"
	"github.com/fioncat/roxide/pkg/choice"
	"github.com/fioncat/roxide/pkg/context"
	"github.com/fioncat/roxide/pkg/db"
	"github.com/fioncat/roxide/pkg/repoutils"
	"github.com/spf13/cobra"
)

func newRepo() *cobra.Command {
	var opts repoOptions

	c := &cobra.Command{
		Use:   "repo [HEAD] [QUERY]",
		Short: "Open the repository in the browser",

		Args: cobra.MaximumNArgs(2),

		ValidArgsFunction: cmd.BuildCompletion(cmd.RepoCompletion),
	}

	return cmd.BuildWithForceNoCache(c, &opts)
}

type repoOptions struct {
	args []string
}

func (o *repoOptions) Complete(c *cobra.Command, args []string) error {
	o.args = args
	return nil
}

func (o *repoOptions) Run(ctx *context.Context) error {
	var repo *db.Repository
	var err error
	if len(o.args) == 0 {
		repo, err = repoutils.GetCurrentRepo(ctx)
		if err != nil {
			return err
		}
	}

	if repo == nil {
		ch := choice.New(ctx, o.args)
		repo, err = ch.One(choice.OneOptions{
			Mode: choice.ModeSelect,
		})
		if err != nil {
			return err
		}
	}

	err = ctx.SetRepo(repo)
	if err != nil {
		return err
	}

	api, err := ctx.RemoteAPI(repo.Remote)
	if err != nil {
		return err
	}

	apiRepo, err := api.GetRepo(repo.Owner, repo.Name)
	if err != nil {
		return err
	}

	return openURL(apiRepo.WebURL)
}

File: src/cmd/open/repo.rs
use std::path::Path;

use anyhow::{Context, Result, bail};
use async_trait::async_trait;
use clap::Args;

use crate::cmd::{CacheArgs, Command, UpstreamArgs, complete};
use crate::config::context::ConfigContext;
use crate::debug;
use crate::exec::git::branch::Branch;
use crate::repo::current::get_current_repo;

/// Open current repository in the browser.
#[derive(Debug, Args)]
pub struct OpenRepoCommand {
    /// By default, opens the repository's root directory. If specified, opens the current
    /// branch of the repository. Can also specify a branch name.
    #[arg(long, short)]
    pub branch: Option<Option<String>>,

    #[clap(flatten)]
    pub cache: CacheArgs,

    #[clap(flatten)]
    pub upstream: UpstreamArgs,
}

#[async_trait]
impl Command for OpenRepoCommand {
    async fn run(self, ctx: ConfigContext) -> Result<()> {
        debug!("[cmd] Run open repo command: {:?}", self);
        ctx.lock()?;

        let repo = get_current_repo(&ctx)?;
        let api = ctx.get_api(&repo.remote, self.cache.force_no_cache)?;

        let mut api_repo = api.get_repo(&repo.remote, &repo.owner, &repo.name).await?;
        if self.upstream.enable {
            let Some(upstream) = api_repo.upstream else {
                bail!("repository {} does not have an upstream", repo.full_name());
            };
            api_repo = api
                .get_repo(&repo.remote, &upstream.owner, &upstream.name)
                .await?;
        }

        let mut web_url = api_repo.web_url;
        if let Some(branch) = self.branch {
            let branch = match branch {
                Some(b) => b,
                None => Branch::current(ctx.git())?,
            };

            let path = Path::new(&web_url).join("tree").join(branch);
            web_url = format!("{}", path.display());
        }

        open::that(&web_url).with_context(|| format!("cannot open repo web url {web_url:?}"))?;
        Ok(())
    }

    fn complete_command() -> clap::Command {
        clap::Command::new("repo").args([complete::branch_arg().short('b')])
    }
}
